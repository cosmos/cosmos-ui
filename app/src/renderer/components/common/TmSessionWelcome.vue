<template>
  <div id="session-welcome" class="tm-session">
    <div class="tm-session-container">
      <div class="tm-session-header">
        <div class="tm-session-title">
          Sign in to Cosmos Voyager
        </div>
        <a @click="closeSession">
          <i class="material-icons">close</i>
        </a>
      </div>

      <div class="tm-session-main">
        <li-session
          icon="usb"
          title="Sign in with Ledger Nano S"
          subtitle="If you have a Ledger Wallet, choose this option."
          @click.native="setState('hardware')"
        />
        <template
          v-if="session.insecureMode"
        >
          <li-session
            v-if="accountExists"
            id="sign-in-with-account"
            icon="lock"
            title="Sign in with password"
            subtitle="If you have an account, choose this option."
            @click.native="setState('sign-in')"
          />
          <li-session
            icon="person_add"
            title="Create new account"
            subtitle="Generate a brand new seed and create a new account."
            @click.native="setState('sign-up')"
          />
          <li-session
            icon="settings_backup_restore"
            title="Import with seed"
            subtitle="Use an existing seed phrase to create an account."
            @click.native="setState('import')"
          />
        </template>
      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters } from "vuex"
import LiSession from "common/TmLiSession"
export default {
  name: `tm-session-welcome`,
  components: {
    LiSession
  },
  computed: {
    ...mapGetters([`session`, `lastPage`]),
    accountExists() {
      return this.session.accounts.length > 0
    }
  },
  methods: {
    setState(value) {
      this.$store.commit(`setSessionModalView`, value)
    },
    closeSession() {
      this.$store.commit(`toggleSessionModal`, false)
    },
  }
}
</script>
