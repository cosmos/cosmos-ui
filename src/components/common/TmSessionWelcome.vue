<template>
  <div id="session-welcome" class="tm-session">
    <div class="tm-session-container">
      <div class="tm-session-header">
        <div class="tm-session-title">
          Sign in to Lunie
        </div>
        <a @click="closeSession">
          <i class="material-icons">close</i>
        </a>
      </div>

      <div class="tm-session-main">
        <LiSession
          icon="usb"
          title="Sign in with Ledger Nano"
          :disabled="!session.browserWithLedgerSupport"
          @click.native="
            () => session.browserWithLedgerSupport && setState('hardware')
          "
        >
          <div
            v-if="session.browserWithLedgerSupport"
            slot="li-session-subtitle"
          >
            If you have a Ledger Wallet, choose this option.<br />
            Don't have a Ledger yet?
            <a
              href="https://shop.ledger.com/?r=3dd204ef7508"
              target="_blank"
              rel="noopener noreferrer"
            >
              Get one here</a
            >.
          </div>

          <div
            v-if="!session.browserWithLedgerSupport"
            slot="li-session-subtitle"
          >
            Please use Chrome, Opera, or Brave. Ledger is not supported in your
            current browser.
          </div>
        </LiSession>
        <LiSession
          id="explore-address"
          icon="search"
          title="Sign in with Address"
          subtitle="If you want to use Lunie with an address
            , choose this option."
          @click.native="setState('explore')"
        />
        <template v-if="session.insecureMode">
          <div class="danger-zone">
            <div class="header">
              <h1>DANGER ZONE</h1>
              <p>
                Never use accounts created in the browser on a real network. You
                could lose all your money.
              </p>
            </div>
            <LiSession
              v-if="accountExists"
              id="sign-in-with-account"
              icon="lock"
              title="Sign in with password"
              subtitle="If you have an account, choose this option."
              @click.native="setState('sign-in')"
            />
            <LiSession
              icon="person_add"
              title="Create new account"
              subtitle="Generate a brand new seed and create a new account."
              @click.native="setState('sign-up')"
            />
            <LiSession
              v-if="session.developmentMode"
              id="import-seed"
              icon="settings_backup_restore"
              title="Import with seed"
              subtitle="Use an existing seed phrase to create an account."
              @click.native="setState('import')"
            />
          </div>
        </template>
      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters } from "vuex"
import LiSession from "common/TmLiSession"
export default {
  name: `tm-session-welcome`,
  components: {
    LiSession
  },
  computed: {
    ...mapGetters([`session`, `lastPage`]),
    accountExists() {
      return this.session.accounts.length > 0
    }
  },
  methods: {
    setState(value) {
      this.$store.commit(`setSessionModalView`, value)
    },
    closeSession() {
      this.$store.commit(`toggleSessionModal`, false)
    }
  }
}
</script>
<style scoped>
.danger-zone {
  border: 1px solid var(--danger);
}

.danger-zone .header {
  color: var(--danger);
  padding: 1rem 1rem 0 1rem;
}
</style>
