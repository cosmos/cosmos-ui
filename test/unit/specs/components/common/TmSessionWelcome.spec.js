import { shallowMount } from "@vue/test-utils"
import TmSessionWelcome from "common/TmSessionWelcome"

describe(`TmSessionWelcome`, () => {
  let wrapper, $store
  const accounts = []

  beforeEach(() => {
    const getters = {
      session: {
        accounts,
        insecureMode: true,
        developmentMode: true,
        browserWithLedgerSupport: null
      },
      lastPage: `/`
    }
    $store = {
      getters,
      commit: jest.fn(),
      dispatch: jest.fn()
    }
    wrapper = shallowMount(TmSessionWelcome, {
      mocks: {
        $store
      },
      stubs: [`router-link`]
    })
  })

  describe(`header buttons`, () => {
    describe(`closes the session modal`, () => {
      it(`without going to prev page`, () => {
        const $store = { commit: jest.fn() }
        const self = {
          back: jest.fn(),
          $store,
          $router: {
            currentRoute: {
              path: `/`
            }
          }
        }
        TmSessionWelcome.methods.closeSession.call(self)
        expect($store.commit).toHaveBeenCalledWith(`toggleSessionModal`, false)
      })

      it(`going back to prev page`, () => {
        const $store = { commit: jest.fn() }
        const self = {
          back: jest.fn(),
          $store,
          $router: {
            currentRoute: {
              path: `/wallet`
            }
          }
        }
        TmSessionWelcome.methods.closeSession.call(self)
        expect($store.commit).toHaveBeenCalledWith(`toggleSessionModal`, false)
      })
    })
  })

  describe(`without accounts`, () => {
    it(`should not show sign-in link since we have no accounts`, () => {
      expect(wrapper.find(`#sign-in-with-account`).exists()).toBe(false)
    })

    it(`has the expected html structure`, () => {
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })

  describe(`with accounts`, () => {
    beforeEach(() => {
      const getters = {
        session: { accounts: [`foo`, `bar`], insecureMode: true },
        lastPage: `/`
      }
      $store = {
        getters,
        commit: jest.fn(),
        dispatch: jest.fn()
      }
      wrapper = shallowMount(TmSessionWelcome, {
        mocks: {
          $store
        }
      })
    })

    it(`should show sign-in link since we have accounts`, () => {
      wrapper.vm.setState = jest.fn()
      expect(wrapper.find(`#sign-in-with-account`).exists()).toBe(true)
      wrapper.find(`#sign-in-with-account`).trigger(`click`)
      expect(wrapper.vm.setState).toHaveBeenCalledWith(`sign-in`)
    })

    it(`sets desired login method`, () => {
      Object.defineProperty(window.navigator, `userAgent`, {
        value: `Chrome`,
        writable: true
      })
      wrapper.vm.setState(`xxx`)
      expect($store.commit).toHaveBeenCalledWith(`setSessionModalView`, `xxx`)
    })

    it(`has the expected html structure`, () => {
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })

  describe(`production`, () => {
    it(`should hide sign in with account if users do not opt in`, () => {
      wrapper.vm.session.accounts = [
        {
          name: `test`
        }
      ]
      expect(wrapper.find(`#sign-in-with-account`).exists()).toBe(true)
      wrapper.vm.session.insecureMode = false
      expect(wrapper.find(`#sign-in-with-account`).exists()).toBe(false)
    })

    it(`should hide seed import if not in development`, () => {
      expect(wrapper.find(`#import-seed`).exists()).toBe(true)
      wrapper.vm.session.developmentMode = false
      expect(wrapper.find(`#import-seed`).exists()).toBe(false)
    })
  })
})
