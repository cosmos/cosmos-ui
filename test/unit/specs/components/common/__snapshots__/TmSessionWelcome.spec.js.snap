// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionWelcome with accounts has the expected html structure 1`] = `
<div
  class="session"
  id="session-welcome"
>
  <a>
    <i
      class="material-icons session-close"
    >
      close
    </i>
  </a>
   
  <div
    class="session-header"
  >
    <img
      class="lunie-logo"
      src="~assets/images/cosmos-wallet-logo.svg"
    />
     
    <h2
      class="session-title"
    >
      
      Welcome to Lunie!
    
    </h2>
     
    <p
      class="session-paragraph"
    >
      
      Lunie is the cryptocurrency wallet for the new staking economy. Easily
      stake your ATOMs, withdraw your rewards, and participate in governance.
    
    </p>
  </div>
   
  <div
    class="session-list"
  >
    <lisession-stub
      icon="person_add"
      title="Create a new address"
    />
     
    <lisession-stub
      icon="person"
      title="Use an existing address"
    />
  </div>
   
  <p
    class="footnote"
  >
    
    By using Lunie, you accept our
    
    <router-link
      class="link"
      to="/terms"
    >
      Terms of Service
    </router-link>
     and
    
    <router-link
      class="link"
      to="/privacy"
    >
      Privacy Policy
    </router-link>
    .
  
  </p>
</div>
`;

exports[`TmSessionWelcome without accounts has the expected html structure 1`] = `
<div
  class="session"
  id="session-welcome"
>
  <a>
    <i
      class="material-icons session-close"
    >
      close
    </i>
  </a>
   
  <div
    class="session-header"
  >
    <img
      class="lunie-logo"
      src="~assets/images/cosmos-wallet-logo.svg"
    />
     
    <h2
      class="session-title"
    >
      
      Welcome to Lunie!
    
    </h2>
     
    <p
      class="session-paragraph"
    >
      
      Lunie is the cryptocurrency wallet for the new staking economy. Easily
      stake your ATOMs, withdraw your rewards, and participate in governance.
    
    </p>
  </div>
   
  <div
    class="session-list"
  >
    <lisession-stub
      icon="person_add"
      title="Create a new address"
    />
     
    <lisession-stub
      icon="person"
      title="Use an existing address"
    />
  </div>
   
  <p
    class="footnote"
  >
    
    By using Lunie, you accept our
    
    <router-link-stub
      class="link"
      to="/terms"
    >
      Terms of Service
    </router-link-stub>
     and
    
    <router-link-stub
      class="link"
      to="/privacy"
    >
      Privacy Policy
    </router-link-stub>
    .
  
  </p>
</div>
`;
